ports/espressif/bindings/espnow/Communicate.c

```diff
@@ -0,0 +1,201 @@
+//|
+//|         This blocks until a timeout of ``2`` seconds if the ESP-NOW internal buffers are full.
+//|
+//|         :param ReadableBuffer message: The message to send (length <= 250 bytes).
+//|         :param Peer peer: Send message to this peer. If `None`, send to all registered peers.
+//|         """
+//|         ...
+STATIC mp_obj_t espnow_com___send(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_message, ARG_peer };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_message,  MP_ARG_OBJ | MP_ARG_REQUIRED },
+        { MP_QSTR_peer,     MP_ARG_OBJ, { .u_obj = mp_const_none } },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    espnow_obj_t *self = pos_args[0];
+    common_hal_espnow_check_for_deinit(self);
+
+    // Get a pointer to the data buffer of the message
+    mp_buffer_info_t message;
+    mp_get_buffer_raise(args[ARG_message].u_obj, &message, MP_BUFFER_READ);
+
+    const uint8_t *mac = NULL;
+    if (args[ARG_peer].u_obj != mp_const_none) {
+        const espnow_peer_obj_t *peer = MP_OBJ_FROM_PTR(mp_arg_validate_type_or_none(args[ARG_peer].u_obj, &espnow_peer_type, MP_QSTR_peer));
+        mac = peer->peer_info.peer_addr;
+    }
+
+    return common_hal_espnow_send(self, &message, mac);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(espnow_com___send_obj, 2, espnow_com___send);
+
+//|     def __read(self) -> Optional[ESPNowPacket]:
+//|         """Read a packet from the receive buffer.
+//|
+//|         This is non-blocking, the packet is received asynchronously from the peer(s).
+//|
+//|         :returns: An `ESPNowPacket` if available in the buffer, otherwise `None`."""
+//|         ...
+STATIC mp_obj_t espnow_com___read(mp_obj_t self_in) {
+    espnow_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_espnow_check_for_deinit(self);
+
+    return common_hal_espnow_read(self);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(espnow_com___read_obj, espnow_com___read);
+
+STATIC const mp_rom_map_elem_t espnow_com_locals_dict_table[] = {
+    // Config parameters
+    { MP_ROM_QSTR(MP_QSTR_job),     MP_ROM_PTR(&espnow_com_job_obj) },
+
+    // Packet statistics
+    { MP_ROM_QSTR(MP_QSTR_success), MP_ROM_PTR(&espnow_com_success_obj) },
+    { MP_ROM_QSTR(MP_QSTR_failure), MP_ROM_PTR(&espnow_com_failure_obj) },
+
+    // Communication methods
+    { MP_ROM_QSTR(MP_QSTR___send),  MP_ROM_PTR(&espnow_com___send_obj) },
+    { MP_ROM_QSTR(MP_QSTR___read),  MP_ROM_PTR(&espnow_com___read_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(espnow_com_locals_dict, espnow_com_locals_dict_table);
+
+//|     def __call__(self, *args: Optional[Any], **kwargs: Optional[Any]) -> Optional[Any]:
+//|         """Calls the private `__send` or `__read` methods with the supplied ``args`` and ``kwargs``
+//|         based on whether the job parameter is set to ``send`` or ``read``."""
+//|         ...
+//|
```

Comment:
I do not understand the reason for this special method and the private `__send()` and `__read()` methods, instead of just `send()` and `read()`.